src/js/components/modals/ServiceFormModal.js

```diff
@@ -0,0 +1,240 @@
+      jsonMode: false,
+      model,
+      service: ServiceUtil.createServiceFromFormModel(model)
+    });
+  }
+
+  handleClearError() {
+    this.setState({
+      errorMessage: null
+    });
+  }
+
+  handleJSONChange(app) {
+    let {service} = this.state;
+    try {
+      service = new Service(JSON.parse(app));
+    } catch (e) {
+
+    }
+    this.setState({app, service})
+  }
+
+  handleJSONToggle() {
+    let nextState = {};
+    if (!this.state.jsonMode) {
+      let {model} = this.triggerAdvancedSubmit();
+      let service = ServiceUtil.createServiceFromFormModel(model);
+      nextState.model = model;
+      nextState.service = service;
+      nextState.app = JSON.stringify(ServiceUtil
+        .getAppDefinitionFromService(service), null, 2);
+    }
+    nextState.jsonMode = !this.state.jsonMode;
+    this.setState(nextState);
+  }
+
+  onMarathonStoreServiceCreateSuccess() {
+    this.resetState();
+    this.props.onClose();
+  }
+
+  onMarathonStoreServiceCreateError(errorMessage) {
+    this.setState({
+      errorMessage
+    });
+  }
+
+  handleCancel() {
+    this.props.onClose();
+  }
+
+  handleSubmit() {
+    if (this.state.jsonMode) {
+      MarathonStore.createService(JSON.parse(this.state.app));
+      return;
+    }
+    if (this.triggerAdvancedSubmit) {
+      let {model} = this.triggerAdvancedSubmit();
+      let service = ServiceUtil.createServiceFromFormModel(model);
+      this.setState({
+        service: service,
+        model: model,
+        errorMessage: null
+      });
+      MarathonStore
+        .createService(ServiceUtil.getAppDefinitionFromService(service));
+    }
+  }
+
+  getAdvancedSubmit(triggerSubmit) {
+    this.triggerAdvancedSubmit = triggerSubmit;
+  }
+
+  getErrorMessage() {
+    let {errorMessage} = this.state;
+    if (!errorMessage) {
+      return null;
+    }
+
+    return (
+      <div>
+        <h4 className="text-align-center text-danger flush-top">
+          {errorMessage.message}
+        </h4>
+        <pre className="text-danger">
+          {JSON.stringify(errorMessage.details, null, 2)}
+        </pre>
+        <button
+          className="button button-small button-danger flush-bottom"
+          onClick={this.handleClearError}>
+          clear
+        </button>
+      </div>
+    );
+  }
+
+  getFooter() {
+    return (
+      <div className="button-collection flush-bottom">
+        <button
+          className="button button-large flush-top flush-bottom"
+          onClick={this.handleCancel}>
+          Cancel
+        </button>
+        <ToggleButton
+          checked={this.state.jsonMode}
+          onChange={this.handleJSONToggle}>
+          JSON mode
+        </ToggleButton>
+        <button
+          className="button button-large button-success flush-bottom"
+          onClick={this.handleSubmit}>
+          Deploy
+        </button>
+      </div>
+    );
+  }
+
+  getModalContents() {
+    let {app, jsonMode, service} = this.state;
+    if (jsonMode) {
```

Comment:
Still have the feeling the this should be part of another component. But that's another PR.